samcli/hook_packages/terraform/hooks/prepare.py

```diff
@@ -0,0 +1,379 @@
+"""
+Terraform prepare hook implementation
+"""
+
+from dataclasses import dataclass
+import json
+import os
+from subprocess import run, CalledProcessError
+from tempfile import NamedTemporaryFile
+from typing import Any, Callable, Dict, List, Optional
+import hashlib
+import logging
+
+from samcli.lib.hook.exceptions import PrepareHookException
+from samcli.lib.utils.hash import str_checksum
+from samcli.lib.utils.resources import (
+    AWS_LAMBDA_FUNCTION as CFN_AWS_LAMBDA_FUNCTION,
+)
+
+LOG = logging.getLogger(__name__)
+
+TF_AWS_LAMBDA_FUNCTION = "aws_lambda_function"
+PROVIDER_NAME = "registry.terraform.io/hashicorp/aws"
+
+# max logical id len is 255
+LOGICAL_ID_HASH_LEN = 8
+LOGICAL_ID_MAX_HUMAN_LEN = 247
+
+PropertyBuilder = Callable[[dict], Any]
+PropertyBuilderMapping = Dict[str, PropertyBuilder]
+
+
+@dataclass
+class ResourceTranslator:
+    cfn_name: str
+    property_builder_mapping: PropertyBuilderMapping
+
+
+def prepare(params: dict) -> dict:
+    """
+    Prepares a terraform application for use with the SAM CLI
+
+    Parameters
+    ----------
+    params: dict
+        Parameters of the IaC application
+
+    Returns
+    -------
+    dict
+        information of the generated metadata files
+    """
+    output_dir_path = params.get("OutputDirPath")
+    if not output_dir_path:
+        raise PrepareHookException("OutputDirPath was not supplied")
+
+    try:
+        # initialize terraform application
+        LOG.info("Initializing Terraform application")
+        run(["terraform", "init"], check=True, capture_output=True)
+
+        # get json output of terraform plan
+        LOG.info("Creating terraform plan and getting JSON output")
+        with NamedTemporaryFile() as temp_file:
+            run(["terraform", "plan", "-out", temp_file.name], check=True, capture_output=True)
+            result = run(["terraform", "show", "-json", temp_file.name], check=True, capture_output=True)
+        tf_json = json.loads(result.stdout)
+
+        # convert terraform to cloudformation
+        LOG.info("Generating metadata file")
+        cfn_dict = _translate_to_cfn(tf_json)
+
+        # store in supplied output dir
+        if not os.path.exists(output_dir_path):
+            os.mkdir(output_dir_path)
+        metadataFilePath = os.path.join(output_dir_path, "template.json")
+        LOG.info("Finished generating metadata file. Storing in %s", metadataFilePath)
+        with open(metadataFilePath, "w+") as metadata_file:
+            json.dump(cfn_dict, metadata_file)
+
+        return {"iac_applications": {"MainApplication": {"metadata_file": metadataFilePath}}}
+
+    except CalledProcessError as e:
+        # one of the subprocess.run calls resulted in non-zero exit code or some OS error
+        LOG.debug(
```

Comment:
Do we need to verify if the terraform binary exists? and prompt to install if it does not?